### DISCUSSION TITLE
Play .MOD soundtracks tunes on your naked LGT8F328p

### DISCUSSION THREAD

**Discussion by SuperUserNameMan** [dbuezas/lgt8fx#105] [Show and tell]
Hello there !

Last october 2020, I've managed to port [pocketmod](https://github.com/rombankzero/pocketmod) to the LGT8F328p.

I did not touch the code since then, but I wanted to share it with you here : [lgt8_mod_player]( https://github.com/SuperUserNameMan/lgt8_mod_player ).

It can play .MOD music format with up to 4 channels and 15 ~~or 31~~ instruments at up to ~12kHz soundrate.

If you don't know what .MOD are : they are tune/music format from the 80's that were initially created using the Amiga. [ see wikipedia ]( https://en.wikipedia.org/wiki/MOD_(file_format) ). They use audio samples as instruments, and they are still used today for demo scene and some game engines.

The music will be outputed through the DAC0 pin of the LGT8. So may want to connect it to an audio-amplifier.

---
UPdates :
- 2021-04-14 : added FastPWM audio mode (also works on real ATmega328p)
- 2021-04-13 : cleaned the code for better PROGMEM compatibility 

---

@prosper00  posted a [video showing the LGT8 playing a MOD](https://github.com/dbuezas/lgt8fx/discussions/105#discussioncomment-577800)

**Comment by prosper00** [comment]
Whoa, cool! I remember being amazed by mod files back in the 90s, and it occurred to me once or twice that a modern 8 bit micro ought to be able to make some good sounds. Will definitely check it out

**Reply by prosper00** [reply]
I've also found DAC0 to have a seriously high output impedance. Higher than your average audio amp can deal with. I see significant distortion with a load much less than a megohm or two.

Definitely recommend a opamp follower when usin the DAC. The mcp6002 would seem fairly ideal, as it can do rail to rail input and output

**Reply by SuperUserNameMan** [reply]
Since I'm not very knowledgeable in op-amp and analog electronics in general, I managed to output something listenable using a cheap PAM8403 delta audio amplifier module somehow hooked to an headphone through a voltage divider ... indeed it was far from perfect : just enough to debug and enjoy the challenge of playing a MOD on this cheap chip.

I don't remember if I actually tested that, but maybe we could use a high speed PWM output instead of the DAC ?

**Reply by prosper00** [reply]
Yup, high speed pwm would work, though it needs to be quite high speed in order to eliminate the need for analog filtering. The pam8403 you used has a pwm frequency of 260kHz as an example.

An op amp circuit should be relatively trivial - just a simple voltage follower configuration. I'll build something up next week, and sketch a schematic

**Comment by prosper00** [comment]
Yeah, a simple follower works great. On the MCP6002, connect pins 1 and 2 together, feed your signal into pin 3, and take your output from pin 1/2. +5V to pin 8, GND to pin 4.

Looking at the signal output on DAC0, it doesn't ever seem to cover the full range from 0V to 5V. Which means that we don't NEED a rail to rail opamp. The super common and cheap LM358 would probably work just as well in this particular application. It's main weakness is that it can't swing all the way up to 5V. It also has a slow slew rate (like the 6002), but we're not slewing very far or very fast, so I think it's adequate.


Cool little project. I'll have to see if I can find a way to convert some more mod files.




![Screenshot at 2021-04-05 10-53-29](https://user-images.githubusercontent.com/57373245/113600478-4f9ebe00-95fd-11eb-8f5b-890cbd958d56.png)

Oddly enough, this seems to clip at about 3V, even though I'm running at 5V. This is measured directly from the DAC pin before any amplification or buffering:
![328mod](https://user-images.githubusercontent.com/57373245/113600484-52011800-95fd-11eb-971e-2ee3e0b55ed9.png)

**Reply by prosper00** [reply]
Wait - R1 should go from pin 2 of the pot to gnd, not pin1 to gnd. In any case, it's optional, all it does is adjust the response of the (linear) pot to something a little more logarithmic.

**Reply by SuperUserNameMan** [reply]
> I'll have to see if I can find a way to convert some more mod files.

Unfortunately, I don't remember how I managed to do so ... I could not find evidence of any script, so it must be a linux command line that I used.
But as you probably already noticed by yourself, it's just a matter of converting the .MOD binary file to a C array.

> Oddly enough, this seems to clip at about 3V, even though I'm running at 5V. This is measured directly from the DAC pin before any amplification or buffering:

Maybe a bug that I did not notice ?
The sketch crashed sometimes too, depending on the compilation optimization level, debug mode and .MOD.

> The super common and cheap LM358 would probably work just as well in this particular application.

I found an MC1747CP2 and a LM741, but they both need symmetrical power supply ... I'll try my chance and see they work in the 0-5V range.
( Oh, i just found two LM386 ... )
edit : they don't seems to work on 5V ... or maybe they are damaged.

**Reply by SuperUserNameMan** [reply]
> Oddly enough, this seems to clip at about 3V, even though I'm running at 5V. This is measured directly from the DAC pin before any amplification or buffering:

On my oscillo (when playing the default module), the signal from DAC does not appears clamped at all. 
Its range is from ~1V to ~4V with an average offset around 2.4V.

EDIT : apparently, these values will vary a little depending on actual Vcc (when powered through USB, Vcc is 4.5V).

**Reply by prosper00** [reply]
> Unfortunately, I don't remember how I managed to do so ... I could not find evidence of any script, so it must be a linux command line that I used.
> But as you probably already noticed by yourself, it's just a matter of converting the .MOD binary file to a C array.

it's 'xxd -i modfile.mod > modfile.h'

>  The sketch crashed sometimes too, depending on the compilation optimization level, debug mode and .MOD.

It is interesting. I have noticed the pitch changes with optimization level. Optimization is basically voodoo to my brain, so I have no idea what to make of that

**Reply by SuperUserNameMan** [reply]
> It is interesting. I have noticed the pitch changes with optimization level. 

I was going to tell you that the pitch change was because of the adaptive sound-rate, but then I noticed that I did not finish to implement the adaptive sound-rate ... So, I have no explanation. Sorry.

**Reply by SuperUserNameMan** [reply]
> I have noticed the pitch changes with optimization level. Optimization is basically voodoo to my brain, so I have no idea what to make of that

Ok now I remember why : it's just that lower optimization levels requires more time to calculate each samples.
If it exceeds ``AUDIO_DT``, the audio will be distorted and slower, thus lowering the tone.

At 32MHz, with a 4 channel .MOD and the fastest compilation optimization level, the optimum sample rate is 12500 Hz.
And the ``AUDIO_DT`` is 80µs.

**Reply by prosper00** [reply]
Makes sense. Too bad that aggressive optimization increases the size, though.

**Comment by SuperUserNameMan** [comment]
Something funny I noticed as I wanted to record a video with audio of the LGT8 playing a tune :

the PWM sound from the  PAM8403 delta audio amplifier module i used to feed a speaker can't be recorded on my smartphone.

I can hear the audio with my naked ears, but none of the microphone from my smartphone can record it ....

**Reply by prosper00** [reply]
https://youtu.be/nhpTtFZ4WVE

This is using a tinny 1" speaker and a TDA7052A amplifier, buit straight out of its datasheet

**Reply by SuperUserNameMan** [reply]
great ! :-D

I'm going to add a link to your comment into the top message.

**Comment by SuperUserNameMan** [comment]
I added Fast PWM audio output ( 62kHz PWM @ 32MHz ).

It also works on real ATmega328p (31kHz PWM @ 16MHz).
